src/runtime/hexagon_remote/elf.h

```diff
@@ -0,0 +1,901 @@
+                mask = 63 << 20;
+
+            } else if ((inst >> 24) == 72) {
+                // Example instruction encoding that has this high byte (ignoring bits 1 and 2):
+                // 0100 1ii0  000i iiii  PPit tttt  iiii iiii
+                if (debug) log_printf("Instruction-specific case A\n");
+                mask = 0x061f20ff;
+            } else if ((inst >> 24) == 73) {
+                // 0100 1ii1  000i iiii  PPii iiii  iiid dddd
+                if (debug) log_printf("Instruction-specific case B\n");
+                mask = 0x061f3fe0;
+            } else if ((inst >> 24) == 120) {
+                // 0111 1000  ii-i iiii  PPii iiii  iiid dddd
+                if (debug) log_printf("Instruction-specific case C\n");
+                mask = 0x00df3fe0;
+            } else if ((inst >> 16) == 27209) {
+                // 0110 1010  0100 1001  PP-i iiii  i--d dddd
+                mask = 0x00001f80;
+            } else if ((inst >> 25) == 72) {
+                // 1001 0ii0  101s ssss  PPii iiii  iiid dddd
+                // 1001 0ii1  000s ssss  PPii iiii  iiid dddd
+                mask = 0x06003fe0;
+
+            } else {
+                log_printf("Unhandled!\n");
+                fail(__LINE__);
+                return;
+            }
+        }
+
+        uintptr_t old_val = val;
+        bool consumed_every_bit = false;
+        for (int i = 0; i < 32; i++) {
+            if (mask & (1 << i)) {
+                if (inst & (1 << i)) {
+                    // This bit should be zero in the unrelocated instruction
+                    fail(__LINE__);
+                    return;
+                }
+                // Consume a bit of val
+                int next_bit = val & 1;
+                if (is_signed) {
+                    consumed_every_bit |= ((intptr_t)val) == -1;
+                    val = ((intptr_t)val) >> 1;
+                } else {
+                    val = ((uintptr_t)val) >> 1;
+                }
+                consumed_every_bit |= (val == 0);
+                inst |= (next_bit << i);
+            }
+        }
+
+        if (verify && !consumed_every_bit) {
+            fail(__LINE__);
+            log_printf("Relocation overflow inst=%08lx mask=%08lx val=%08lx\n", inst, mask, (unsigned long)old_val);
+            return;
+        }
+
+        if (debug) log_printf("Relocated instruction:\n  %08lx\n", inst);
+        *((uint32_t *)addr) = inst;
+    }
+
+    // Do all the relocations for sec (e.g. .text), using the list of
+    // relocations in sec_rela (e.g. .rela.text)
+    void do_relocations_for_section(section_header_t *sec, section_header_t *sec_rela) {
+        if (!sec_rela || !sec) {
+            fail(__LINE__);
+            return;
+        }
+
+        struct known_sym {
+            const char *name;
+            char *addr;
+        };
+        static known_sym known_syms[] = {
+            {"close", (char *)(&close)},
```

Comment:
I don't understand why dlsym(NULL, ...) doesn't find these... add a comment explaining that it doesn't. If you know why, that would be nice to know :)